api/pkg/cmd/stack.go

```diff
@@ -1,65 +1,61 @@
 package cmd
 
 import (
+	"context"
+	"encoding/json"
+
 	"github.com/chanzuckerberg/happy/api/pkg/dbutil"
 	"github.com/chanzuckerberg/happy/shared/model"
 	"github.com/pkg/errors"
-	"gorm.io/gorm/clause"
+	"github.com/sirupsen/logrus"
 )
 
-type Stack interface {
-	CreateOrUpdateAppStack(model.AppStackPayload) (*model.AppStack, error)
-	GetAppStacks(model.AppStackPayload) ([]*model.AppStack, error)
-	DeleteAppStack(model.AppStackPayload) (*model.AppStack, error)
+type StackIface interface {
+	GetAppStacks(context.Context, model.AppStackPayload) ([]*model.AppStack, error)
+	// CreateOrUpdateAppStack(model.AppStackPayload) (*model.AppStack, error)
+	// DeleteAppStack(model.AppStackPayload) (*model.AppStack, error)
 }
 
-type dbStack struct {
-	DB *dbutil.DB
+type Stack struct {
+	db  StackIface
+	ecs StackIface
+	eks StackIface
 }
 
-func MakeStack(db *dbutil.DB) Stack {
-	return &dbStack{
-		DB: db,
+func MakeStack(db *dbutil.DB) StackIface {
+	return &Stack{
+		// DB is not currently used since this is currently just a read interface for the old data locations
+		// but we should keep this here so it's easy to set up later when we want to move the data
+		db:  MakeStackBackendDB(db),
+		ecs: &StackBackendECS{},
+		eks: &StackBackendEKS{},
 	}
 }
 
-func (s *dbStack) CreateOrUpdateAppStack(payload model.AppStackPayload) (*model.AppStack, error) {
-	db := s.DB.GetDB()
-	stack := &model.AppStack{AppStackPayload: payload}
-	res := db.Clauses(clause.OnConflict{
-		Columns: []clause.Column{
-			{Name: "app_name"},
-			{Name: "environment"},
-			{Name: "stack"},
-		},
-		UpdateAll: true,
-	}).Create(&stack)
-
-	return stack, errors.Wrapf(res.Error, "unable to create app stack %s", payload.AppMetadata)
-}
-
-func (s *dbStack) GetAppStacks(payload model.AppStackPayload) ([]*model.AppStack, error) {
-	db := s.DB.GetDB()
-	stack := &model.AppStack{AppStackPayload: payload}
-	stacks := []*model.AppStack{}
-	res := db.Where(stack).Find(&stacks)
-	return stacks, errors.Wrapf(res.Error, "unable to get app stacks for %s", stack.AppMetadata)
+func (s Stack) GetAppStacks(ctx context.Context, payload model.AppStackPayload) ([]*model.AppStack, error) {
+	switch payload.TaskLaunchType {
```

Comment:
alternatively, in MakeStack, do the switch case and instead of having a struct with all three (db, eks, ecs), only return the stack interface that matches.

Reply:
`MakeStack` runs at application boot time but the switch needs to execute per request